File: android/src/main/java/com/rndocumentscanner/utils/DocumentOcrAnalyzer.kt
package com.rndocumentscanner.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.margelo.nitro.NitroModules
import com.margelo.nitro.rndocumentscanner.ExtractedDocumentData
import java.io.File

object DocumentOcrAnalyzer {

    fun extractData(imageUri: String, documentType: String): ExtractedDocumentData {
        val bitmap = loadImage(imageUri) ?: return emptyResult()
        val rotated = rotateImageIfRequired(bitmap, imageUri)

        try {
            val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
            val inputImage = InputImage.fromBitmap(rotated, 0)
            val visionText: Text = Tasks.await(recognizer.process(inputImage))

            val allLines = mutableListOf<String>()
            for (block in visionText.textBlocks) {
                for (line in block.lines) {
                    val trimmed = line.text.trim()
                    if (trimmed.isNotEmpty()) {
                        allLines.add(trimmed)
                    }
                }
            }

            return if (documentType == "passport") {
                parsePassport(allLines)
            } else {
                parseCccd(allLines)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return emptyResult()
        } finally {
            if (!rotated.isRecycled) {
                rotated.recycle()
            }
            if (bitmap != rotated && !bitmap.isRecycled) {
                bitmap.recycle()
            }
        }
    }

    private fun parseCccd(lines: List<String>): ExtractedDocumentData {
        var idNumber = ""
        var fullName = ""
        var dateOfBirth = ""
        var gender = ""
        var nationality = "Việt Nam"
        var placeOfOrigin = ""
        var placeOfResidence = ""
        var expiryDate = ""
        var issueDate = ""

        // Regex patterns
        val idRegex = Regex("""\b(\d{12})\b""")
        val oldIdRegex = Regex("""\b(\d{9})\b""")
        val dateRegex = Regex("""\b(\d{2}[/.-]\d{2}[/.-]\d{4})\b""")

        for (i in lines.indices) {
            val line = lines[i]
            val lower = line.lowercase()

            // 1. Số CCCD (12 chữ số)
            if (idNumber.isEmpty()) {
                val match = idRegex.find(line) ?: oldIdRegex.find(line)
                if (match != null) {
                    idNumber = match.value
                } else if (lower.contains("số") || lower.contains("no")) {
                    val digits = line.filter { it.isDigit() }
                    if (digits.length in 9..12) {
                        idNumber = digits
                    }
                }
            }

            // 2. Họ và tên
            if (fullName.isEmpty()) {
                if (lower.contains("họ và tên") || lower.contains("full name") || lower.contains("họ tên")) {
                    val cleaned = line.replace(Regex("""(?i)(họ và tên|full name|họ tên|[:;.,-])"""), "").trim()
                    if (cleaned.length >= 3 && cleaned.any { it.isLetter() }) {
                        fullName = cleaned
                    } else if (i + 1 < lines.size) {
                        val nextLine = lines[i + 1].trim()
                        if (nextLine.length >= 3 && isUppercaseName(nextLine)) {
                            fullName = nextLine
                        }
                    }
                } else if (isUppercaseName(line) && !lower.contains("cộng hòa") && !lower.contains("độc lập") && !lower.contains("căn cước") && !lower.contains("việt nam")) {
                    fullName = line
                }
            }

            // 3. Ngày sinh
            if (dateOfBirth.isEmpty()) {
                if (lower.contains("ngày sinh") || lower.contains("date of birth") || lower.contains("sinh ngày") || lower.contains("dob")) {
                    val match = dateRegex.find(line)
                    if (match != null) {
                        dateOfBirth = match.value
                    } else if (i + 1 < lines.size) {
                        val nextMatch = dateRegex.find(lines[i + 1])
                        if (nextMatch != null) dateOfBirth = nextMatch.value
                    }
                } else {
                    val match = dateRegex.find(line)
                    if (match != null && dateOfBirth.isEmpty()) {
                        dateOfBirth = match.value
                    }
                }
            }

            // 4. Giới tính
            if (gender.isEmpty()) {
                if (lower.contains("nam") && !lower.contains("việt nam")) {
                    gender = "Nam"
                } else if (lower.contains("nữ") || lower.contains("nu")) {
                    gender = "Nữ"
                }
            }

            // 5. Quốc tịch
            if (lower.contains("quốc tịch") || lower.contains("nationality")) {
                if (lower.contains("việt nam") || lower.contains("viet nam")) {
                    nationality = "Việt Nam"
                }
            }

            // 6. Quê quán
            if (placeOfOrigin.isEmpty() && (lower.contains("quê quán") || lower.contains("place of origin"))) {
                val cleaned = line.replace(Regex("""(?i)(quê quán|place of origin|[:;])"""), "").trim()
                if (cleaned.isNotEmpty()) {
                    placeOfOrigin = cleaned
                } else if (i + 1 < lines.size) {
                    placeOfOrigin = lines[i + 1].trim()
                }
            }

            // 7. Nơi thường trú
            if (placeOfResidence.isEmpty() && (lower.contains("thường trú") || lower.contains("place of residence") || lower.contains("nơi thường trú"))) {
                val cleaned = line.replace(Regex("""(?i)(nơi thường trú|thường trú|place of residence|[:;])"""), "").trim()
                if (cleaned.isNotEmpty()) {
                    placeOfResidence = cleaned
                } else if (i + 1 < lines.size) {
                    placeOfResidence = lines[i + 1].trim()
                }
            }

            // 8. Hạn sử dụng
            if (expiryDate.isEmpty() && (lower.contains("giá trị đến") || lower.contains("expiry") || lower.contains("có giá trị"))) {
                val match = dateRegex.find(line)
                if (match != null) {
                    expiryDate = match.value
                } else if (lower.contains("không thời hạn") || lower.contains("vô thời hạn")) {
                    expiryDate = "Không thời hạn"
                }
            }
        }

        return ExtractedDocumentData(
            idNumber = idNumber.ifEmpty { "Chưa nhận diện được số" },
            fullName = fullName.ifEmpty { "Chưa nhận diện được họ tên" },
            dateOfBirth = dateOfBirth.ifEmpty { "Chưa rõ" },
            gender = gender.ifEmpty { "Chưa rõ" },
            nationality = nationality,
            placeOfOrigin = placeOfOrigin.ifEmpty { "Chưa rõ" },
            placeOfResidence = placeOfResidence.ifEmpty { "Chưa rõ" },
            expiryDate = expiryDate.ifEmpty { "Theo quy định" },
            issueDate = issueDate.ifEmpty { "Theo quy định" },
            mrzLines = arrayOf(),
            rawText = lines.toTypedArray()
        )
    }

    private fun parsePassport(lines: List<String>): ExtractedDocumentData {
        var idNumber = ""
        var fullName = ""
        var dateOfBirth = ""
        var gender = ""
        var nationality = "Việt Nam"
        var expiryDate = ""
        val mrzLines = mutableListOf<String>()

        // 1. Tìm các dòng MRZ chuẩn ICAO Doc 9303 (Bắt đầu với P< hoặc có nhiều dấu '<')
        for (line in lines) {
            val clean = line.replace(" ", "").uppercase()
            if (clean.startsWith("P<") || clean.startsWith("P<<") || (clean.length >= 30 && clean.contains("<<"))) {
                mrzLines.add(clean)
            } else if (clean.length >= 30 && clean.contains("<") && clean.any { it.isDigit() }) {
                mrzLines.add(clean)
            }
        }

        // Bóc tách từ MRZ nếu tìm thấy đủ 2 dòng
        if (mrzLines.size >= 2) {
            val line1 = mrzLines[0]
            val line2 = mrzLines[1]

            // Line 1: P<VNMNGUYEN<<VAN<AN<<<<...
            try {
                if (line1.length >= 5) {
                    nationality = if (line1.substring(2, 5) == "VNM") "Việt Nam" else line1.substring(2, 5)
                    val namesPart = line1.substring(5).replace("<", " ").trim()
                    fullName = namesPart.split("  ").joinToString(" ")
                }
            } catch (_: Exception) {}

            // Line 2: C1234567<8VNM9808154M3308159001098012345<<68
            try {
                if (line2.length >= 20) {
                    val passNum = line2.substring(0, 9).replace("<", "")
                    idNumber = passNum

                    // DOB (YYMMDD)
                    val dobRaw = line2.substring(13, 19)
                    if (dobRaw.length == 6 && dobRaw.all { it.isDigit() }) {
                        val yy = dobRaw.substring(0, 2)
                        val mm = dobRaw.substring(2, 4)
                        val dd = dobRaw.substring(4, 6)
                        val yearPrefix = if (yy.toInt() > 30) "19" else "20"
                        dateOfBirth = "$dd/$mm/$yearPrefix$yy"
                    }

                    // Sex (M/F)
                    val sexChar = line2[20]
                    gender = when (sexChar) {
                        'M' -> "Nam"
                        'F' -> "Nữ"
                        else -> "Chưa rõ"
                    }

                    // Expiry (YYMMDD)
                    val expRaw = line2.substring(21, 27)
                    if (expRaw.length == 6 && expRaw.all { it.isDigit() }) {
                        val yy = expRaw.substring(0, 2)
                        val mm = expRaw.substring(2, 4)
                        val dd = expRaw.substring(4, 6)
                        expiryDate = "$dd/$mm/20$yy"
                    }
                }
            } catch (_: Exception) {}
        }

        // Nếu MRZ chưa đủ, fallback regex từ các dòng chữ thường
        if (idNumber.isEmpty()) {
            val passportRegex = Regex("""\b([A-Z]\d{7,8})\b""")
            for (line in lines) {
                val match = passportRegex.find(line)
                if (match != null) {
                    idNumber = match.value
                    break
                }
            }
        }

        if (fullName.isEmpty()) {
            for (line in lines) {
                val lower = line.lowercase()
                if (isUppercaseName(line) && !lower.contains("passport") && !lower.contains("hộ chiếu") && !lower.contains("việt nam") && !lower.contains("cộng hòa")) {
                    fullName = line
                    break
                }
            }
        }

        return ExtractedDocumentData(
            idNumber = idNumber.ifEmpty { "Chưa rõ số hộ chiếu" },
            fullName = fullName.ifEmpty { "Chưa rõ họ tên" },
            dateOfBirth = dateOfBirth.ifEmpty { "Chưa rõ" },
            gender = gender.ifEmpty { "Chưa rõ" },
            nationality = nationality,
            placeOfOrigin = "",
            placeOfResidence = "",
            expiryDate = expiryDate.ifEmpty { "Theo quy định" },
            issueDate = "Theo quy định",
            mrzLines = mrzLines.toTypedArray(),
            rawText = lines.toTypedArray()
        )
    }

    private fun isUppercaseName(str: String): Boolean {
        val trimmed = str.trim()
        val words = trimmed.split(" ").filter { it.isNotEmpty() }
        if (words.size !in 2..6) return false
        return words.all { word ->
            word.all { it.isLetter() && (it.isUpperCase() || !it.isLowerCase()) }
        }
    }

    private fun emptyResult(): ExtractedDocumentData {
        return ExtractedDocumentData(
            idNumber = "Chưa nhận diện được",
            fullName = "Chưa nhận diện được",
            dateOfBirth = "Chưa rõ",
            gender = "Chưa rõ",
            nationality = "Việt Nam",
            placeOfOrigin = "Chưa rõ",
            placeOfResidence = "Chưa rõ",
            expiryDate = "Chưa rõ",
            issueDate = "Chưa rõ",
            mrzLines = arrayOf(),
            rawText = arrayOf()
        )
    }

    private fun loadImage(uriString: String): Bitmap? {
        val clean = if (uriString.startsWith("file://")) uriString.substring(7) else uriString
        val file = File(clean)
        if (file.exists()) {
            return BitmapFactory.decodeFile(file.absolutePath)
        }
        try {
            val uri = Uri.parse(uriString)
            val context = NitroModules.applicationContext
            if (context != null && uri.scheme == "content") {
                context.contentResolver.openInputStream(uri)?.use { stream ->
                    return BitmapFactory.decodeStream(stream)
                }
            }
        } catch (_: Exception) {}
        return null
    }

    private fun rotateImageIfRequired(img: Bitmap, path: String): Bitmap {
        val clean = if (path.startsWith("file://")) path.substring(7) else path
        val file = File(clean)
        if (!file.exists()) return img

        return try {
            val ei = ExifInterface(clean)
            val orientation = ei.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
            val degree = when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_90 -> 90f
                ExifInterface.ORIENTATION_ROTATE_180 -> 180f
                ExifInterface.ORIENTATION_ROTATE_270 -> 270f
                else -> 0f
            }
            if (degree != 0f) {
                val matrix = Matrix()
                matrix.postRotate(degree)
                Bitmap.createBitmap(img, 0, 0, img.width, img.height, matrix, true)
            } else {
                img
            }
        } catch (_: Exception) {
            img
        }
    }
}

File: ios/Utils/DocumentOcrAnalyzer.swift
import Foundation
import UIKit
import Vision
#if canImport(NitroModules)
import NitroModules
#endif

public final class DocumentOcrAnalyzer {

    public static func extractData(imageUri: String, documentType: String) -> ExtractedDocumentData {
        guard let image = loadImage(uriString: imageUri),
              let cgImage = image.cgImage else {
            return emptyResult()
        }

        var detectedLines: [String] = []
        let semaphore = DispatchSemaphore(value: 0)

        let request = VNRecognizeTextRequest { request, error in
            defer { semaphore.signal() }
            guard error == nil,
                  let observations = request.results as? [VNRecognizedTextObservation] else {
                return
            }

            for observation in observations {
                if let topCandidate = observation.topCandidates(1).first {
                    let text = topCandidate.string.trimmingCharacters(in: .whitespacesAndNewlines)
                    if !text.isEmpty {
                        detectedLines.append(text)
                    }
                }
            }
        }

        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true

        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        do {
            try handler.perform([request])
            semaphore.wait()
        } catch {
            return emptyResult()
        }

        if documentType == "passport" {
            return parsePassport(lines: detectedLines)
        } else {
            return parseCccd(lines: detectedLines)
        }
    }

    private static func parseCccd(lines: [String]) -> ExtractedDocumentData {
        var idNumber = ""
        var fullName = ""
        var dateOfBirth = ""
        var gender = ""
        var nationality = "Việt Nam"
        var placeOfOrigin = ""
        var placeOfResidence = ""
        var expiryDate = ""
        let issueDate = "Theo quy định"

        let idPattern = "\\b(\\d{12})\\b"
        let oldIdPattern = "\\b(\\d{9})\\b"
        let datePattern = "\\b(\\d{2}[/.-]\\d{2}[/.-]\\d{4})\\b"

        for (index, line) in lines.enumerated() {
            let lower = line.lowercased()

            // 1. Số CCCD
            if idNumber.isEmpty {
                if let match = matchRegex(pattern: idPattern, in: line) ?? matchRegex(pattern: oldIdPattern, in: line) {
                    idNumber = match
                } else if lower.contains("số") || lower.contains("no") {
                    let digits = line.filter { $0.isNumber }
                    if digits.count >= 9 && digits.count <= 12 {
                        idNumber = digits
                    }
                }
            }

            // 2. Họ và tên
            if fullName.isEmpty {
                if lower.contains("họ và tên") || lower.contains("full name") || lower.contains("họ tên") {
                    let cleaned = line.replacingOccurrences(of: "(?i)(họ và tên|full name|họ tên|[:;.,-])", with: "", options: .regularExpression).trimmingCharacters(in: .whitespacesAndNewlines)
                    if cleaned.count >= 3 && cleaned.rangeOfCharacter(from: .letters) != nil {
                        fullName = cleaned
                    } else if index + 1 < lines.count {
                        let nextLine = lines[index + 1].trimmingCharacters(in: .whitespacesAndNewlines)
                        if nextLine.count >= 3 && isUppercaseName(nextLine) {
                            fullName = nextLine
                        }
                    }
                } else if isUppercaseName(line) && !lower.contains("cộng hòa") && !lower.contains("độc lập") && !lower.contains("căn cước") && !lower.contains("việt nam") {
                    fullName = line
                }
            }

            // 3. Ngày sinh
            if dateOfBirth.isEmpty {
                if lower.contains("ngày sinh") || lower.contains("date of birth") || lower.contains("sinh ngày") {
                    if let match = matchRegex(pattern: datePattern, in: line) {
                        dateOfBirth = match
                    } else if index + 1 < lines.count, let nextMatch = matchRegex(pattern: datePattern, in: lines[index + 1]) {
                        dateOfBirth = nextMatch
                    }
                } else if let match = matchRegex(pattern: datePattern, in: line) {
                    dateOfBirth = match
                }
            }

            // 4. Giới tính
            if gender.isEmpty {
                if lower.contains("nam") && !lower.contains("việt nam") {
                    gender = "Nam"
                } else if lower.contains("nữ") || lower.contains("nu") {
                    gender = "Nữ"
                }
            }

            // 5. Quốc tịch
            if lower.contains("quốc tịch") || lower.contains("nationality") {
                if lower.contains("việt nam") || lower.contains("viet nam") {
                    nationality = "Việt Nam"
                }
            }

            // 6. Quê quán
            if placeOfOrigin.isEmpty && (lower.contains("quê quán") || lower.contains("place of origin")) {
                let cleaned = line.replacingOccurrences(of: "(?i)(quê quán|place of origin|[:;])", with: "", options: .regularExpression).trimmingCharacters(in: .whitespacesAndNewlines)
                if !cleaned.isEmpty {
                    placeOfOrigin = cleaned
                } else if index + 1 < lines.count {
                    placeOfOrigin = lines[index + 1].trimmingCharacters(in: .whitespacesAndNewlines)
                }
            }

            // 7. Nơi thường trú
            if placeOfResidence.isEmpty && (lower.contains("thường trú") || lower.contains("place of residence") || lower.contains("nơi thường trú")) {
                let cleaned = line.replacingOccurrences(of: "(?i)(nơi thường trú|thường trú|place of residence|[:;])", with: "", options: .regularExpression).trimmingCharacters(in: .whitespacesAndNewlines)
                if !cleaned.isEmpty {
                    placeOfResidence = cleaned
                } else if index + 1 < lines.count {
                    placeOfResidence = lines[index + 1].trimmingCharacters(in: .whitespacesAndNewlines)
                }
            }

            // 8. Hạn sử dụng
            if expiryDate.isEmpty && (lower.contains("giá trị đến") || lower.contains("expiry") || lower.contains("có giá trị")) {
                if let match = matchRegex(pattern: datePattern, in: line) {
                    expiryDate = match
                } else if lower.contains("không thời hạn") || lower.contains("vô thời hạn") {
                    expiryDate = "Không thời hạn"
                }
            }
        }

        return ExtractedDocumentData(
            idNumber: idNumber.isEmpty ? "Chưa nhận diện được số" : idNumber,
            fullName: fullName.isEmpty ? "Chưa nhận diện được họ tên" : fullName,
            dateOfBirth: dateOfBirth.isEmpty ? "Chưa rõ" : dateOfBirth,
            gender: gender.isEmpty ? "Chưa rõ" : gender,
            nationality: nationality,
            placeOfOrigin: placeOfOrigin.isEmpty ? "Chưa rõ" : placeOfOrigin,
            placeOfResidence: placeOfResidence.isEmpty ? "Chưa rõ" : placeOfResidence,
            expiryDate: expiryDate.isEmpty ? "Theo quy định" : expiryDate,
            issueDate: issueDate,
            mrzLines: [],
            rawText: lines
        )
    }

    private static func parsePassport(lines: [String]) -> ExtractedDocumentData {
        var idNumber = ""
        var fullName = ""
        var dateOfBirth = ""
        var gender = ""
        var nationality = "Việt Nam"
        var expiryDate = ""
        var mrzLines: [String] = []

        for line in lines {
            let clean = line.replacingOccurrences(of: " ", with: "").uppercased()
            if clean.hasPrefix("P<") || clean.hasPrefix("P<<") || (clean.count >= 30 && clean.contains("<<")) {
                mrzLines.append(clean)
            } else if clean.count >= 30 && clean.contains("<") && clean.rangeOfCharacter(from: .decimalDigits) != nil {
                mrzLines.append(clean)
            }
        }

        if mrzLines.count >= 2 {
            let line1 = mrzLines[0]
            let line2 = mrzLines[1]

            // Line 1
            if line1.count >= 5 {
                let country = String(line1.prefix(5).suffix(3))
                nationality = (country == "VNM") ? "Việt Nam" : country
                let namesPart = String(line1.dropFirst(5)).replacingOccurrences(of: "<", with: " ").trimmingCharacters(in: .whitespacesAndNewlines)
                fullName = namesPart.components(separatedBy: "  ").filter { !$0.isEmpty }.joined(separator: " ")
            }

            // Line 2
            if line2.count >= 20 {
                let passNum = String(line2.prefix(9)).replacingOccurrences(of: "<", with: "")
                idNumber = passNum

                // DOB
                if line2.count >= 19 {
                    let dobIndexStart = line2.index(line2.startIndex, offsetBy: 13)
                    let dobIndexEnd = line2.index(line2.startIndex, offsetBy: 19)
                    let dobRaw = String(line2[dobIndexStart..<dobIndexEnd])
                    if dobRaw.count == 6 && dobRaw.allSatisfy({ $0.isNumber }) {
                        let yy = String(dobRaw.prefix(2))
                        let mm = String(dobRaw.dropFirst(2).prefix(2))
                        let dd = String(dobRaw.suffix(2))
                        let yearPrefix = (Int(yy) ?? 0 > 30) ? "19" : "20"
                        dateOfBirth = "\(dd)/\(mm)/\(yearPrefix)\(yy)"
                    }
                }

                // Sex
                if line2.count >= 21 {
                    let sexIndex = line2.index(line2.startIndex, offsetBy: 20)
                    let sexChar = line2[sexIndex]
                    gender = (sexChar == "M") ? "Nam" : (sexChar == "F" ? "Nữ" : "Chưa rõ")
                }

                // Expiry
                if line2.count >= 27 {
                    let expIndexStart = line2.index(line2.startIndex, offsetBy: 21)
                    let expIndexEnd = line2.index(line2.startIndex, offsetBy: 27)
                    let expRaw = String(line2[expIndexStart..<expIndexEnd])
                    if expRaw.count == 6 && expRaw.allSatisfy({ $0.isNumber }) {
                        let yy = String(expRaw.prefix(2))
                        let mm = String(expRaw.dropFirst(2).prefix(2))
                        let dd = String(expRaw.suffix(2))
                        expiryDate = "\(dd)/\(mm)/20\(yy)"
                    }
                }
            }
        }

        if idNumber.isEmpty {
            for line in lines {
                if let match = matchRegex(pattern: "\\b([A-Z]\\d{7,8})\\b", in: line) {
                    idNumber = match
                    break
                }
            }
        }

        return ExtractedDocumentData(
            idNumber: idNumber.isEmpty ? "Chưa rõ số hộ chiếu" : idNumber,
            fullName: fullName.isEmpty ? "Chưa rõ họ tên" : fullName,
            dateOfBirth: dateOfBirth.isEmpty ? "Chưa rõ" : dateOfBirth,
            gender: gender.isEmpty ? "Chưa rõ" : gender,
            nationality: nationality,
            placeOfOrigin: "",
            placeOfResidence: "",
            expiryDate: expiryDate.isEmpty ? "Theo quy định" : expiryDate,
            issueDate: "Theo quy định",
            mrzLines: mrzLines,
            rawText: lines
        )
    }

    private static func isUppercaseName(_ str: String) -> Bool {
        let words = str.split(separator: " ").filter { !$0.isEmpty }
        if words.count < 2 || words.count > 6 { return false }
        return words.allSatisfy { word in
            word.allSatisfy { $0.isLetter && ($0.isUppercase || !$0.isLowercase) }
        }
    }

    private static func matchRegex(pattern: String, in text: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let nsString = text as NSString
        let results = regex.matches(in: text, range: NSRange(location: 0, length: nsString.length))
        guard let first = results.first else { return nil }
        return nsString.substring(with: first.range)
    }

    private static func emptyResult() -> ExtractedDocumentData {
        return ExtractedDocumentData(
            idNumber: "Chưa nhận diện được",
            fullName: "Chưa nhận diện được",
            dateOfBirth: "Chưa rõ",
            gender: "Chưa rõ",
            nationality: "Việt Nam",
            placeOfOrigin: "Chưa rõ",
            placeOfResidence: "Chưa rõ",
            expiryDate: "Chưa rõ",
            issueDate: "Chưa rõ",
            mrzLines: [],
            rawText: []
        )
    }

    private static func loadImage(uriString: String) -> UIImage? {
        let clean = uriString.replacingOccurrences(of: "file://", with: "")
        if FileManager.default.fileExists(atPath: clean) {
            return UIImage(contentsOfFile: clean)
        }
        if let url = URL(string: uriString), let data = try? Data(contentsOf: url) {
            return UIImage(data: data)
        }
        return nil
    }
}
